src/Storages/PartitionStrategy.cpp

```diff
@@ -0,0 +1,326 @@
+        const Block & sample_block,
+        ContextPtr context)
+    {
+        HiveStylePartitionStrategy::PartitionExpressionActionsAndColumnName actions_with_column_name;
+        ASTs concat_args;
+
+        if (const auto * tuple_function = partition_by->as<ASTFunction>();
+            tuple_function && tuple_function->name == "tuple")
+        {
+            chassert(tuple_function->arguments->children.size() == partition_columns.size());
+
+            std::size_t index = 0;
+
+            for (const auto & partition_column : partition_columns)
+            {
+                const auto & child = tuple_function->arguments->children[index++];
+
+                concat_args.push_back(std::make_shared<ASTLiteral>(partition_column.name + "="));
+
+                concat_args.push_back(makeASTFunction("toString", child));
+
+                concat_args.push_back(std::make_shared<ASTLiteral>("/"));
+            }
+        }
+        else
+        {
+            chassert(partition_columns.size() == 1);
+
+            ASTs to_string_args = {1, partition_by};
+            concat_args.push_back(std::make_shared<ASTLiteral>(partition_columns.front().name + "="));
+            concat_args.push_back(makeASTFunction("toString", std::move(to_string_args)));
+            concat_args.push_back(std::make_shared<ASTLiteral>("/"));
+        }
+
+        ASTPtr hive_expr = makeASTFunction("concat", std::move(concat_args));
+        auto hive_syntax_result = TreeRewriter(context).analyze(hive_expr, sample_block.getNamesAndTypesList());
+        actions_with_column_name.actions = ExpressionAnalyzer(hive_expr, hive_syntax_result, context).getActions(false);
+        actions_with_column_name.column_name = hive_expr->getColumnName();
+
+        return actions_with_column_name;
+    }
+
+    Block buildBlockWithoutPartitionColumns(
+        const Block & sample_block,
+        const std::unordered_set<std::string> & partition_expression_required_columns_set)
+    {
+        Block result;
+        for (size_t i = 0; i < sample_block.columns(); i++)
+        {
+            if (!partition_expression_required_columns_set.contains(sample_block.getByPosition(i).name))
+            {
+                result.insert(sample_block.getByPosition(i));
+            }
+        }
+
+        return result;
+    }
+}
+
+PartitionStrategy::PartitionStrategy(ASTPtr partition_by_, const Block & sample_block_, ContextPtr context_)
+: partition_by(partition_by_), sample_block(sample_block_), context(context_)
+{
+    auto key_description = KeyDescription::getKeyFromAST(partition_by, ColumnsDescription::fromNamesAndTypes(sample_block.getNamesAndTypes()), context);
+    partition_columns = key_description.sample_block.getNamesAndTypesList();
+}
+
+const NamesAndTypesList & PartitionStrategy::getPartitionColumns() const
+{
+    return partition_columns;
+}
+
+struct HivePartitionStrategyFactory
+{
+    static std::shared_ptr<PartitionStrategy> get(
```

Comment:
Frankly speaking this struct with a single static method, which is actually used only in 1 place, looks like an overcomplication. Let's either use just a static method `static std::shared_ptr<PartitionStrategy> createHiveStylePartitionSrategy(...)` or just move this code inside `PartitionStrategyFactory::get` where it used, as this creator code is not really big.